packages/task/src/browser/task-terminal-widget-manager.ts

```diff
@@ -0,0 +1,182 @@
+import { WorkspaceService } from '@theia/workspace/lib/browser/workspace-service';
+
+export interface TaskTerminalWidgetOpenerOptions extends WidgetOpenerOptions, TerminalWidgetFactoryOptions {
+    taskId: number;
+    taskConfig?: TaskConfiguration;
+}
+
+export interface TaskTerminalWidget extends TerminalWidget {
+    readonly kind: 'task';
+    dedicated?: boolean;
+    taskId?: number;
+    taskConfig?: TaskConfiguration;
+    busy?: boolean;
+}
+export namespace TaskTerminalWidget {
+    export function is(widget: TerminalWidget): widget is TaskTerminalWidget {
+        return widget.kind === 'task';
+    }
+}
+
+export namespace TaskTerminalWidgetOpenerOptions {
+    export function isDedicatedTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskConfig && !!options.taskConfig.presentation && options.taskConfig.presentation.panel === PanelKind.Dedicated;
+    }
+
+    export function isNewTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskConfig && !!options.taskConfig.presentation && options.taskConfig.presentation.panel === PanelKind.New;
+    }
+
+    export function isSharedTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskConfig &&
+            (options.taskConfig.presentation === undefined || options.taskConfig.presentation.panel === undefined || options.taskConfig.presentation.panel === PanelKind.Shared);
+    }
+}
+
+@injectable()
+export class TaskTerminalWidgetManager {
+
+    @inject(ApplicationShell)
+    protected readonly shell: ApplicationShell;
+
+    @inject(TaskDefinitionRegistry)
+    protected readonly taskDefinitionRegistry: TaskDefinitionRegistry;
+
+    @inject(TerminalService)
+    protected readonly terminalService: TerminalService;
+
+    @inject(TaskWatcher)
+    protected readonly taskWatcher: TaskWatcher;
+
+    @inject(TaskServer)
+    protected readonly taskServer: TaskServer;
+
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    @postConstruct()
+    protected init(): void {
+        this.taskWatcher.onTaskExit((event: TaskExitedEvent) => {
+            const finishedTaskId = event.taskId;
+            // find the terminal where the task ran, and mark it as "idle"
+            for (const terminal of this.getTaskTerminalWidgets()) {
+                if (terminal.taskId === finishedTaskId) {
+                    terminal.busy = false;
+                    this.notifyTaskFinished(terminal);
+                    break;
+                }
+            }
+        });
+
+        this.terminalService.onDidCreateTerminal(async (widget: TerminalWidget) => {
+            const terminal = TaskTerminalWidget.is(widget) && widget;
+            if (terminal) {
+                const didConectListener = terminal.onDidOpen(async () => {
+                    const context = this.workspaceService.workspace && this.workspaceService.workspace.uri;
```

Comment:
Should not we dispose `didConnectListener` here? i.e. intermediate reconnections should not affect busy state.

We should test it by:
- creating a new task terminal
- running a long running task
- switching off connection to the server
- trying to run a new long running task
- switching on connection to the server
- the original task terminal should not be used for second task, whether should not be any unexpected effects, after reconnection is reestablished, the output of original long running task should be shown.

It is important the server should not be stopped and started. It is different use case. You can use Gitpod and switch off/no your internet connection to reproduce or some local proxy to drop traffic to localhost.

Reply:
> * switching off connection to the server
> * trying to run a new long running task

sorry one thing confuses me: how am i able to start a new long running task while the connection is off?

Reply:
> Should not we dispose `didConnectListener` here? i.e. intermediate reconnections should not affect busy state.

I tested the scenarios of server reconnects on Gitpod.

Server reconnect works - terminal widget successfully attaches to the process, task output is properly displayed in the terminal widget, and the terminal widget can be used for the next task.
However I believe `didConnectListener` should not be disposed in the callback of `terminal.onDidOpen`. It causes `Terminal will be reused by tasks.` not to be displayed for new tasks that are started after the reconnect.

Reply:
ok, to me most important that you test such cases :) If code is required to make them work let's keep it